plugins/systemd/systemd_mem

```diff
@@ -0,0 +1,121 @@
+#!/bin/sh
+: << =cut
+
+=head1 NAME
+
+systemd_mem - Plugin to graph the memory usage of multiple systemd services as reported by systemctl
+
+=head1 CONFIGURATION
+
+The environment variable "services" is used to provide a space separated list of services to graph.
+
+Example:
+   [systemd_mem]
+   env.services apache2 tomcat9 mysql
+
+env.services defaults to "munin-node".
+
+=head1 VERSION
+
+1.0
+
+=head1 AUTHOR
+
+Paul Alexandrow <paul@alexandrow.org>
+
+=head1 LICENSE
+
+Copyright 2021 Paul Alexandrow, paul@alexandrow.org
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software
+and associated documentation files (the "Software"), to deal in the Software without
+restriction, including without limitation the rights to use, copy, modify, merge, publish,
+distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or
+substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+=head1 MAGIC MARKERS
+
+#%# family=auto
+#%# capabilities=autoconf
+
+=cut
+
+. "$MUNIN_LIBDIR/plugins/plugin.sh"
+
+get_services() {
+    services=${services:-"munin-node"}
+    echo "$services" | awk 'BEGIN{RS=" "}{$1=$1}1'
+}
+
+output_config() {
+    echo "graph_title Systemd Service Memory Usage"
+    echo "graph_info Memory usage for selected services as reported by systemctl"
+    echo "graph_vlabel bytes"
+    echo "graph_args --base 1024 --lower-limit 0"
+    echo "graph_category memory"
+    for service in $(get_services); do
+        clean_name="$(clean_fieldname "$service")"
+        description=$(systemctl show "$service" --all | grep ^Description= | cut -d '=' -f 2)
+        printf "%s.label %s\n" "$clean_name" "$description"
+        printf "%s.info memory usage\n" "$clean_name"
+    done
+}
+
+output_values() {
+    for service in $(get_services); do
+        clean_name="$(clean_fieldname "$service")"
+        usage=$(systemctl show "$service" --all | grep ^MemoryCurrent= | cut -d '=' -f 2)
+        if [ "$usage" = "[not set]" ]; then
+            usage=0
```

Comment:
Maybe you could use `U` here. Munin would interpret this as *unknown* (e.g. due to an unexpected condition).

Reply:
Afaik systemctl returns `[not set]` only in case a service is just not running. So from a logical point of view, it is not necessarily unexpected and also it _is_ known how much memory is used - zero.

And from a practical point of view, I use a lower alerting limit of 1 on some services to get notified when they stopped running. I know that there are better solutions for that, but it's still a convenient side effect.

Reply:
Yes, I agree - zero seems to be a reasonable value in this case.